### ISSUE TITLE
How to Install??

### ISSUE THREAD

**Issue by pGit1** [aleju/imgaug#2]
This looks amazing!! How to install and make use of this library?

**Comment by aleju** [comment]
I didn't have the time to write an installer yet. For now, you will have to copy the files `imgaug.py`, `augmenters.py` and `parameters.py` to your project's directory and import them from there. Then follow the examples from the readme.

**Comment by pGit1** [comment]
So is this a generator? What is mean is does it generate new batches of augmented data during the fitting process? This seems like a much better version of the ImageDataGenerator in Keras.

I would love to see a real example where you are actually fitting a simple model (a neural net preferably).

My other question is whether the sequential class adds each augmentation one after the other OR does it pick one Augmentation per image from the sample space of all possible transformations you specify and only applies that one?

All in all your documentation is great I am just a noob! An example where you are actually fitting a model would be amazing! :D

**Comment by aleju** [comment]
It's not a generator. You call the library with your images (as one numpy array or a list of numpy arrays), the library augments them and then returns the images (again, as one arrays or list of arrays) in one step.

The augmentations are randomly sampled per image. So if you use horizontal flipping with 50% probability and provide 10 images, you expect roughly 5 of them to be flipped.

Without having tested it, the following should roughly be a keras example:
```python
from imgaug import augmenters as iaa
from from keras.layers import Input, Convolution2D, Dense, Activation
from keras.models import Model

NB_IMAGES = 1000
INPUT_HEIGHT = 32
INPUT_WIDTH = 32
NB_BATCHES = 10000
BATCH_SIZE = 128

# generate 1000 random 32x32 images with random yes/no labels
X = np.random.randint(0, 255, (NB_IMAGES, INPUT_HEIGHT, INPUT_WIDTH, 3), dtype=np.uint8)
Y = np.random.binomial(1, 0.5, size=(NB_IMAGES,))

# create a simple keras model
model_input = Input(shape=(INPUT_HEIGHT, INPUT_WIDTH, 3), name="images")
model = Permute((3, 1, 2))(model_input) # move the channel axis to the front (H, W, C) to (C, H, W) 
model = Convolution2D(16, 3, 3)(model)
model = Activation("relu")(model)
model = Flatten()(model)
model = Dense(1)(model)
model_output = Activation("sigmoid")(model)
model = Model(input=[model_input], output=[model_output])
model.compile(optimizer="adam", loss="binary_crossentropy")

# define the augmentations
seq = iaa.Sequential([
    iaa.Crop(px=(0, 0.1)), # crop images from each side by 0 to 10%
    iaa.Fliplr(0.5), # horizontally flip 50% of the images
    iaa.GaussianBlur(sigma=(0, 3.0)) # blur images with a sigma of 0 to 3.0
])

# train in a loop on 10k batches
for batch_idx in range(NB_BATCHES):
    # collect randomly chosen images with their labels for the batch
    batch_indices = np.random.randint(0, X.shape[0], (BATCH_SIZE,))
    X_batch = X[batch_indices]
    Y_batch = Y[batch_indices]
    # augment the batch's images
    X_batch_aug = seq.augment_images(X_batch)
    # normalize the images from value range 0-255 to 0-1.0,
    # then train on the batch
    model.train_on_batch(X_batch_aug / 255.0, Y_batch)
```

**Comment by pGit1** [comment]
Alex thank you for the amazing examples and explanation! This is a huge help!!

**Comment by pGit1** [comment]
Alex quick follow up question: 

You say:
> The augmentations are randomly sampled per image.

in an answer to one of my questions above. So in the example code you provide above 

> # define the augmentations
seq = iaa.Sequential([
    iaa.Crop(percent=(0, 0.1)), # crop images from each side by 0 to 10%
    iaa.Fliplr(0.5), # horizontally flip 50% of the images
    iaa.GaussianBlur(sigma=(0, 3.0)) # blur images with a sigma of 0 to 3.0
])

does this mean that seq.augment_image will crop ever image it is fed, flip every other image, and blur every image it is fed in that sequence? 

The root of my question is if I want every image in an array to have drop out alone and gaussian blur alone do I have to create a `seq_droput` and `seq_augmentation` setup separately if I dont want my images to have dropout and then sequentially blurred? So for one image I would get two augmentaions instead of one augmentation where dropout and gaussian blur were added to the same image. Am I thinking of this correctly?

**Comment by aleju** [comment]
The example listed will do the following to each image:
1. Crop each axis of the image by 0% to 10%, e.g. an image may be cropped by 0.1% at the top, 2% at the right, 0.7% at the bottom and 5% at the left.
2. Flip the image with 50% probability.
3. Blur the image with a gaussian kernel that has a sigma between 0 and 3.0, e.g. 1.3.
The values are resampled (i.e. different) for each image (and each batch).

"if I want every image in an array to have drop out alone and gaussian blur alone"
Do you mean that you want to apply either dropout or gaussian blur, but never both (to the same image)? Then you can use `iaa.Sometimes`, which takes a probability `p` and two augmenters/sequentials `A` and `B` and with probability p it applies `A` and otherwise `B`:
seq = iaa.Sequential([
iaa.Fliplr(0.5),
iaa.Sometimes(0.5, iaa.GaussianBlur(sigma=(0, 3.0), iaa.Dropout(0.05))) # either blur or dropout
])

The library doesn't use generators in the sense that it uses nowhere python's `yield` command. The example posted however samples one batch and augments only that batch, so the memory footprint should be the same.

**Comment by pGit1** [comment]
Alex thanks for the response. What you describe above is one use case that I have been experimenting with from the readme, but I think after doing this experimentation what I really want is a little different. 

So what I really want after exploring iaa.Sequence output is for multiple individual augmentations per image. So if I had a file of 5 images, I would want to apply 3 different augmentations per image resulting in 15 augmented images. So for image_1 if I apply Gaussian blur, Add, and Contrast Normalization I should end up with 3 new image_1's (one Gauss augmented, one Add augmented, and one Contrast Normalized). Then proceed to image 2 and do the same thing until we get to the fifth image.

Wondering if there is a way to do this natively with Sequence module or should I just use for loops and call the Augmenters the amount of times I want an augmented copy of an original image. Thanks again!!

**Comment by aleju** [comment]
Currently the easiest way to do that would probably be to define the augmenters as a normal list and loop over that list for each image. Note that each augmenter also has the `augment_images()` function, so you don't have to wrap each of them in a Sequential.

**Comment by pGit1** [comment]
Excellent thanks Alex! Could you provide some example code it doesn't even have to be tested just to get an idea? I this this makes sense though. Would some thing like below be what you are referring to?

```

img_array = some_array_of_images
images_aug = []
aug_lis = [iaa.GaussianBlur((0, 1.5)), 
                iaa.AdditiveGaussianNoise(loc=0, scale=(0.0, 0.2)),
                iaa.Dropout((0.0, 0.1))]

for img in img_array:
    for aug in aug_lis:
        images_aug.append(aug.augment_images(img))
```

**Comment by aleju** [comment]
Yes, that's what I meant. But as you augment a single image each time you have to use either `augment_image(img)` or `augment_images([img])`.

**Comment by pGit1** [comment]
Excellent Thanks!!
